sdk/digitaltwins/Azure.DigitalTwins.Core/src/Queries/QueryBuilders/Linq/ExpressionNormalizer.cs

```diff
@@ -0,0 +1,275 @@
+            NormalizerRewrites = normalizerRewrites;
+        }
+
+        internal Dictionary<Expression, Expression> NormalizerRewrites { get; }
+
+        internal static Expression Normalize(Expression expression, Dictionary<Expression, Expression> rewrites)
+        {
+            Debug.Assert(expression != null, "expression != null");
+            Debug.Assert(rewrites != null, "rewrites != null");
+
+            ExpressionNormalizer normalizer = new ExpressionNormalizer(rewrites);
+            Expression result = normalizer.Visit(expression);
+            return result;
+        }
+
+        internal override Expression VisitBinary(BinaryExpression b)
+        {
+            BinaryExpression visited = (BinaryExpression)base.VisitBinary(b);
+
+            switch (visited.NodeType)
+            {
+                case ExpressionType.Equal:
+                case ExpressionType.NotEqual:
+                case ExpressionType.LessThan:
+                case ExpressionType.LessThanOrEqual:
+                case ExpressionType.GreaterThan:
+                case ExpressionType.GreaterThanOrEqual:
+
+                    Expression normalizedLeft = UnwrapObjectConvert(visited.Left);
+                    Expression normalizedRight = UnwrapObjectConvert(visited.Right);
+                    if (normalizedLeft != visited.Left || normalizedRight != visited.Right)
+                    {
+                        visited = CreateRelationalOperator(visited.NodeType, normalizedLeft, normalizedRight);
+                    }
+                    break;
+            }
+
+            bool leftVisisted = _patterns.TryGetValue(visited.Left, out Pattern pattern);
+            if (leftVisisted && pattern.Kind == PatternKind.Compare && IsConstantZero(visited.Right))
+            {
+                ComparePattern comparePattern = (ComparePattern)pattern;
+                if (TryCreateRelationalOperator(visited.NodeType, comparePattern.Left, comparePattern.Right, out BinaryExpression relationalExpression))
+                {
+                    visited = relationalExpression;
+                }
+            }
+
+            RecordRewrite(b, visited);
+
+            return visited;
+        }
+
+        internal override Expression VisitUnary(UnaryExpression u)
+        {
+            UnaryExpression visited = (UnaryExpression)base.VisitUnary(u);
+            Expression result = visited;
+
+            RecordRewrite(u, result);
+
+            return result;
+        }
+
+        internal override Expression VisitMethodCall(MethodCallExpression call)
+        {
+            Expression visited = VisitMethodCallNoRewrite(call);
+            RecordRewrite(call, visited);
+            return visited;
+        }
+
+        internal Expression VisitMethodCallNoRewrite(MethodCallExpression call)
+        {
+            MethodCallExpression visited = (MethodCallExpression)base.VisitMethodCall(call);
+
+            if (visited.Method.IsStatic
```

Comment:
What if visited is null?

Reply:
It won't be in the way it's used here, but it's not a terrible idea to use your `Ensure` or `Debug.Assert` methods to validate this in case any future maintainers change something:

```C#
            MethodCallExpression visited = base.VisitMethodCall(call) as MethodCallExpression;
            Debug.Assert(visited != null);

            if (visited.Method.IsStatic
```